src/services/smartSelection.ts

```diff
@@ -0,0 +1,266 @@
+                    if (isTemplateSpan(parentNode) && nextNode && isTemplateMiddleOrTemplateTail(nextNode)) {
+                        const start = node.getFullStart() - "${".length;
+                        const end = nextNode.getStart() + "}".length;
+                        pushSelectionRange(start, end);
+                    }
+
+                    // Blocks with braces, brackets, parens, or JSX tags on separate lines should be
+                    // selected from open to close, including whitespace but not including the braces/etc. themselves.
+                    const isBetweenMultiLineBookends = isSyntaxList(node)
+                        && isListOpener(prevNode)
+                        && isListCloser(nextNode)
+                        && !positionsAreOnSameLine(prevNode.getStart(), nextNode.getStart(), sourceFile);
+                    const jsDocCommentStart = hasJSDocNodes(node) && node.jsDoc![0].getStart();
+                    const start = isBetweenMultiLineBookends ? prevNode.getEnd() : node.getStart();
+                    const end = isBetweenMultiLineBookends ? nextNode.getStart() : node.getEnd();
+                    if (isNumber(jsDocCommentStart)) {
+                        pushSelectionRange(jsDocCommentStart, end);
+                    }
+                    pushSelectionRange(start, end);
+
+                    // String literals should have a stop both inside and outside their quotes.
+                    if (isStringLiteral(node) || isTemplateLiteral(node)) {
+                        pushSelectionRange(start + 1, end - 1);
+                    }
+
+                    parentNode = node;
+                    break;
+                }
+            }
+        }
+
+        return selectionRange;
+
+        function pushSelectionRange(start: number, end: number): void {
+            // Skip empty ranges
+            if (start !== end) {
+                // Skip ranges that are identical to the parent
+                const textSpan = createTextSpanFromBounds(start, end);
+                if (!selectionRange || !textSpansEqual(textSpan, selectionRange.textSpan)) {
+                    selectionRange = { textSpan, ...selectionRange && { parent: selectionRange } };
+                }
+            }
+        }
+    }
+
+    /**
+     * Like `ts.positionBelongsToNode`, except positions immediately after nodes
+     * count too, unless that position belongs to the next node. In effect, makes
+     * selections able to snap to preceding tokens when the cursor is on the tail
+     * end of them with only whitespace ahead.
+     * @param pos The position to check.
+     * @param node The candidate node to snap to.
+     * @param nextNode The next sibling node in the tree.
+     * @param sourceFile The source file containing the nodes.
+     */
+    function positionShouldSnapToNode(pos: number, node: Node, nextNode: Node | undefined) {
+        // Can’t use 'ts.positionBelongsToNode()' here because it cleverly accounts

```

Comment:
Why do these files have smart quotes?

Reply:
I habitually [type proper apostrophes by hand](https://twitter.com/atcb/status/1119075745446256640) when writing English sentences, including in code comments. (OS-level quote replacement tends to be unreliable, and hard to fight when it’s wrong, e.g. figures like `50° 32' 24.1"` or `6' 2"` require straight hatches. Bad typography drives me bananas so at some point I disabled all the auto-replacement features and started typing the character I mean.)

Is there some environment where this causes a problem? I’ve never had any issues, nor has anyone ever noticed or pointed it out 😄

Reply:
There's probably nothing wrong with it, but it does feel unusual. I've had issues where one team hands off an HTML file with content that was copy/pasted from Word and that caused issues for one reason or another. So I guess I'm scarred from that.

Reply:
Don’t worry; you can count on me never to paste content from Word into the TypeScript repo 😛